test/e2e/apibinding/apibinding_permissionclaims_by_name_test.go

```diff
@@ -0,0 +1,287 @@
+	require.NotNil(t, binding)
+
+	apiExportVWCfg := rest.CopyConfig(cfg)
+	//nolint:staticcheck // SA1019 VirtualWorkspaces is deprecated but not removed yet
+	apiExportVWCfg.Host = sheriffExport.Status.VirtualWorkspaces[0].URL
+	t.Logf("vwHost: %s", apiExportVWCfg.Host)
+
+	apiExportClient, err := kcpkubernetesclientset.NewForConfig(apiExportVWCfg)
+	require.NoError(t, err)
+
+	t.Logf("verify we can create a new configmap in consumer namespace via the virtual workspace")
+	cm := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "confmap1",
+		},
+	}
+	framework.Eventually(t, func() (done bool, str string) {
+		cm, err = apiExportClient.Cluster(consumerWorkspace.Path()).CoreV1().ConfigMaps(consumerNS1.Name).Create(ctx, cm, metav1.CreateOptions{})
+		if err != nil {
+			return false, err.Error()
+		}
+
+		return true, ""
+	}, wait.ForeverTestTimeout, 110*time.Millisecond, "timed out trying to create configmap in consumer namespace")
+	require.Equal(t, "consumer-ns-1", cm.Namespace)
+	t.Logf("cluster for CM: %s", logicalcluster.From(cm).String())
+
+	t.Logf("verify we can update a configmap in consumer workspace via the virtual workspace")
+	cm.Data = map[string]string{
+		"something": "new",
+	}
+	framework.Eventually(t, func() (done bool, str string) {
+		cm, err = apiExportClient.Cluster(consumerWorkspace.Path()).CoreV1().ConfigMaps(consumerNS1.Name).Update(ctx, cm, metav1.UpdateOptions{})
+		if err != nil {
+			return false, err.Error()
+		}
+
+		return true, ""
+	}, wait.ForeverTestTimeout, 110*time.Millisecond, "timed out trying to update configmap in consumer namespace")
+	require.Equal(t, cm.Data["something"], "new")
+
+	t.Logf("ensure that configmaps in an unspecified namespace cannot be created")
+
+	t.Logf("Creating unclaimed consumer namespace")
+	consumerNS2, err := kubeClusterClient.Cluster(consumerWorkspace.Path()).CoreV1().Namespaces().Create(ctx, &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "consumer-ns-2",
+		},
+	}, metav1.CreateOptions{})
+	require.NoError(t, err, "failed to create %s", consumerNS2.Name)
+
+	t.Logf("Waiting for namespace to exist")
+	framework.Eventually(t, func() (done bool, str string) {
+		consumerNS1, err := kubeClusterClient.Cluster(consumerWorkspace.Path()).CoreV1().Namespaces().Get(ctx, consumerNS2.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err.Error()
+		}
+
+		if consumerNS1.Status.Phase == v1.NamespaceActive {
+			return true, ""
+		}
+
+		return false, "not done waiting for ns2 to be created"
+	}, wait.ForeverTestTimeout, 110*time.Millisecond, "could not wait for namespace to be ready")
+	t.Logf("Namespace %s ready", consumerNS2.Name)
+
+	cm = &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "confmap2",
+		},
+	}
+	framework.Eventually(t, func() (done bool, str string) {
+		// TODO(nrb): This create should be forbidden. However, constructing a URL manually allows creation
```

Comment:
@s-urbaniak I'd like some help here; after talking to @ncdc this appears to be something we'll need to handle within authorization. The current code in the PR handles the case where a namespace/name is claimed, but it does not deny actions where the namespace/name is not claimed.

As an example with kubectl, I have this Permission Claim applied (it was created by this test):

```yaml
❯ k --user shard-admin --server https://192.168.1.245:6443/clusters/c7q9xh6qxb7bup8j get apiexports/wild.wild.west -o yaml
apiVersion: apis.kcp.io/v1alpha1
kind: APIExport
metadata:
  annotations:
    kcp.io/cluster: c7q9xh6qxb7bup8j
    kcp.io/path: root:e2e-workspace-jmfvb:e2e-workspace-9g5ts
  creationTimestamp: "2023-01-06T20:49:18Z"
  generation: 3
  name: wild.wild.west
  resourceVersion: "9418"
  uid: b7692154-ca12-42ac-9032-308509c7769c
spec:
  identity:
    secretRef:
      name: wild.wild.west
      namespace: kcp-system
  latestResourceSchemas:
  - today.sheriffs.wild.wild.west
  permissionClaims:
  - group: ""
    resource: configmaps
    resourceSelector:
    - namespace: consumer-ns-1
status:
  conditions:
  - lastTransitionTime: "2023-01-06T20:49:18Z"
    status: "True"
    type: IdentityValid
  - lastTransitionTime: "2023-01-06T20:49:18Z"
    status: "True"
    type: VirtualWorkspaceURLsReady
  identityHash: 789663fe83a7e14ff8ddb20dc5505f35b02df456b04625f9ed5a453fa5515362
  virtualWorkspaces:
  - url: https://192.168.1.245:6443/services/apiexport/c7q9xh6qxb7bup8j/wild.wild.west
```

Based on that, I should only be able to create a configmap in `consumer-ns-1`. However, creating one via the export's URL and namespace `consumer-ns-2` succeeds.


```shell
☸ workspace.kcp.io/current in kcp on  permclaims-by-name [$!+?⇕] via 🐹 v1.19.4
❮ k --user shard-admin --server https://192.168.1.245:6443/services/apiexport/c7q9xh6qxb7bup8j/wild.wild.west/clusters/31uc19q3o6ozaxvm/ -n consumer-ns-2 create cm "confmap3"
configmap/confmap3 created
```

LIST requests work as I expect, however; no configmaps are exposed when querying `consumer-ns-2`.

```shell
☸ workspace.kcp.io/current in kcp on  permclaims-by-name [$!+?⇕] via 🐹 v1.19.4
❯ k --user shard-admin --server https://192.168.1.245:6443/services/apiexport/c7q9xh6qxb7bup8j/wild.wild.west/clusters/31uc19q3o6ozaxvm/ -n consumer-ns-2 get configmaps
No resources found in consumer-ns-2 namespace.
```

Same with GET requests for a single configmap that we just made.

```shell
☸ workspace.kcp.io/current in kcp on  permclaims-by-name [$?] via 🐹 v1.19.4
❯ k --user shard-admin --server https://192.168.1.245:6443/services/apiexport/c7q9xh6qxb7bup8j/wild.wild.west/clusters/31uc19q3o6ozaxvm/ -n consumer-ns-2 get configmaps/confmap3
Error from server (NotFound): the server could not find the requested resource
```

Reply:
@nrb @ncdc i have coded checking permission claims authorization in the reverse permission claims PR (see https://github.com/kcp-dev/kcp/pull/2089/files#diff-77ec3c56601a3a2c83fab874126c91a8f56d6d0acb35a808bc7cf515edefd546). What we can do:

1. factor the permission claims authorizer minus the checks for claimed verbs.
2. we'll wait until the refined version of reverse permission claims lands.

I am leaning towards 1., taking the permission claims authorizer from that PR should be straight forward and the "only" change would be to skip the claimed verbs check and simply allow access if the resource selector matches.

wdyt?

Reply:
Option 1 seems good to me. I'll try to take that code and rework it so that it uses the resource selector.

Reply:
@s-urbaniak That works too!

Reply:
https://github.com/kcp-dev/kcp/pull/2584